packager/packager.cc

```diff
@@ -1542,9 +1542,82 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
     return info;
 }
 
+void validateLayering(const core::Context &ctx, const Import &i) {
+    if (i.type == ImportType::Test) {
+        return;
+    }
+
+    const auto &packageDB = ctx.state.packageDB();
+    ENFORCE(packageDB.getPackageInfo(i.name.mangledName).exists())
+    ENFORCE(packageDB.getPackageForFile(ctx, ctx.file).exists())
+    auto &thisPkg = PackageInfoImpl::from(packageDB.getPackageForFile(ctx, ctx.file));
+    auto &otherPkg = PackageInfoImpl::from(packageDB.getPackageInfo(i.name.mangledName));
+
+    if (!thisPkg.strictDependenciesLevel.has_value() || !otherPkg.strictDependenciesLevel.has_value() ||
+        !thisPkg.layer.has_value() || !otherPkg.layer.has_value()) {
+        return;
+    }
+
+    if (thisPkg.strictDependenciesLevel.value().first == core::packages::StrictDependenciesLevel::False) {
+        return;
+    }
+
+    auto possibleLayers = packageDB.layers();
+    auto pkgLayer = thisPkg.layer.value().first;
+    auto otherPkgLayer = otherPkg.layer.value().first;
+    auto pkgLayerIndex = packageDB.layerIndex(pkgLayer);
+    auto otherPkgLayerIndex = packageDB.layerIndex(otherPkgLayer);
+
+    if (pkgLayerIndex < otherPkgLayerIndex) {
+        if (auto e = ctx.beginError(i.name.loc, core::errors::Packager::LayeringViolation)) {
+            e.setHeader("`{}` is at layer `{}`, so it can not import package `{}`, which is at layer `{}`",
+                        thisPkg.show(ctx), pkgLayer.show(ctx), otherPkg.show(ctx), otherPkgLayer.show(ctx));
+            e.addErrorLine(core::Loc(thisPkg.loc.file(), thisPkg.layer.value().second), "`{}`'s `{}` declared here",
+                           thisPkg.show(ctx), "layer");
+            e.addErrorLine(core::Loc(otherPkg.loc.file(), otherPkg.layer.value().second), "`{}`'s `{}` declared here",
+                           otherPkg.show(ctx), "layer");
+        }
+    }
+
+    if (otherPkg.strictDependenciesLevel.value().first == core::packages::StrictDependenciesLevel::False) {
+        if (auto e = ctx.beginError(i.name.loc, core::errors::Packager::LayeringViolation)) {
```

Comment:
The docs at http://go/layers only describe the previous error as a "layering violation"

In my mind, this error ("All deps must be layered or higher") is a strict deps violation, not a layering violation.

What do you think about using a different error code?

I'm asking because I'm wondering if it's also worth mentioning in the error header something about strict dependencies specifically (right now the error only mentions layers, but the place where people will find more context about this error is on the docs for strict dependencies, not for layers.

Update: it looks like you've made this change already on your stacked PR, is that right?

Reply:
Yup, when extending the error to handle `Dag`, I realized it was really a strict deps violation. I'll bring that change over to this branch as well, and update the message to say something strict deps related.